File: api/Enkl.Api/Services/ColumnService.cs
using Enkl.Api.Data;
using Enkl.Api.Domain.Entities;
using Enkl.Api.Dtos;
using Microsoft.EntityFrameworkCore;

namespace Enkl.Api.Services;

public class ColumnService
{
    private readonly AppDbContext _db;

    public ColumnService(AppDbContext db)
    {
        _db = db;
    }

    public async Task<ColumnDto> CreateAsync(Guid projectId, CreateColumnRequest request)
    {
        var nextOrder = await _db.Columns.Where(c => c.ProjectId == projectId).CountAsync();
        var column = new Column
        {
            Id = Guid.NewGuid(),
            ProjectId = projectId,
            Name = request.Name,
            Done = request.Done,
            Color = request.Color,
            ColorBackground = request.ColorBackground,
            Order = nextOrder
        };
        _db.Columns.Add(column);
        await _db.SaveChangesAsync();
        return new ColumnDto(column.Id, column.Name, column.Done, column.Color, column.ColorBackground, column.Order, column.Cap);
    }

    public async Task<ColumnDto?> UpdateAsync(Guid projectId, Guid columnId, UpdateColumnRequest request)
    {
        var column = await _db.Columns.FirstOrDefaultAsync(c => c.Id == columnId && c.ProjectId == projectId);
        if (column is null) return null;

        column.Name = request.Name;
        column.Done = request.Done;
        column.Color = request.Color;
        column.ColorBackground = request.ColorBackground;
        column.Order = request.Order;
        // -1 means uncapped; anything <1 (0, other negatives) normalizes back to -1 rather than
        // being rejected — there's no such thing as a column that holds zero tasks — matching
        // clampColumnCap's client-side twin (storage.js).
        column.Cap = request.Cap < 1 ? -1 : request.Cap;
        await _db.SaveChangesAsync();
        return new ColumnDto(column.Id, column.Name, column.Done, column.Color, column.ColorBackground, column.Order, column.Cap);
    }

    public async Task<bool> DeleteAsync(Guid projectId, Guid columnId)
    {
        var column = await _db.Columns.FirstOrDefaultAsync(c => c.Id == columnId && c.ProjectId == projectId);
        if (column is null) return false;

        // TaskItem.ColumnId is a Restrict FK (see TaskItemConfiguration) — deleting a column that
        // still holds tasks would otherwise fail at the DB level. Mirrors mutations.js's deleteColumn:
        // every task in the column is deleted outright (not reassigned elsewhere), with the same
        // cleanup TasksController.Delete does per task — sub-tasks orphaned back to top-level (their
        // ParentTaskId FK is also Restrict), dependency rows removed on both sides since
        // TaskDependency.DependsOnTaskId is Restrict too. Document/Risk/Decision.TaskId and
        // TaskItem.AssigneeId are already SetNull, so those clear themselves.
        var taskIds = await _db.Tasks.Where(t => t.ColumnId == columnId).Select(t => t.Id).ToListAsync();
        if (taskIds.Count > 0)
        {
            var orphanedChildren = await _db.Tasks.Where(t => t.ParentTaskId != null && taskIds.Contains(t.ParentTaskId!.Value) && !taskIds.Contains(t.Id)).ToListAsync();
            foreach (var child in orphanedChildren) child.ParentTaskId = null;

            var dependencies = await _db.TaskDependencies.Where(d => taskIds.Contains(d.TaskId) || taskIds.Contains(d.DependsOnTaskId)).ToListAsync();
            _db.TaskDependencies.RemoveRange(dependencies);

            var tasks = await _db.Tasks.Where(t => taskIds.Contains(t.Id)).ToListAsync();
            _db.Tasks.RemoveRange(tasks);
        }

        _db.Columns.Remove(column);
        await _db.SaveChangesAsync();
        return true;
    }
}

File: mariadb-api/src/Services/ColumnService.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

use Enkl\Api\Support\Uuid;
use PDO;

/** Ported from Services/ColumnService.cs. */
final class ColumnService
{
    public function __construct(private readonly PDO $db)
    {
    }

    public function create(string $projectId, array $request): array
    {
        $stmt = $this->db->prepare('SELECT COUNT(*) FROM "Columns" WHERE "ProjectId" = :pid');
        $stmt->execute(['pid' => $projectId]);
        $nextOrder = (int) $stmt->fetchColumn();

        $id = Uuid::v4();
        $done = (bool) ($request['done'] ?? false);
        $colorBackground = (bool) ($request['colorBackground'] ?? true);
        $stmt = $this->db->prepare(<<<SQL
            INSERT INTO "Columns" ("Id", "ProjectId", "Name", "Done", "Color", "ColorBackground", "Order")
            VALUES (:id, :pid, :name, :done, :color, :colorBackground, :order)
        SQL);
        // PDO's array-form execute() binds every value as PDO::PARAM_STR, and PHP's (string) cast of
        // false is '' — which Postgres's boolean parser rejects (it needs '0'/'1'/'true'/'false'), so
        // bool params must be sent as int here; the DTO below keeps the real PHP bool for json_encode.
        $stmt->execute([
            'id' => $id, 'pid' => $projectId, 'name' => $request['name'] ?? '',
            'done' => (int) $done, 'color' => $request['color'] ?? null, 'colorBackground' => (int) $colorBackground, 'order' => $nextOrder,
        ]);

        return ['id' => $id, 'name' => $request['name'] ?? '', 'done' => $done, 'color' => $request['color'] ?? null, 'colorBackground' => $colorBackground, 'order' => $nextOrder, 'cap' => -1];
    }

    public function update(string $projectId, string $columnId, array $request): ?array
    {
        $stmt = $this->db->prepare('SELECT 1 FROM "Columns" WHERE "Id" = :id AND "ProjectId" = :pid');
        $stmt->execute(['id' => $columnId, 'pid' => $projectId]);
        if ($stmt->fetch() === false) {
            return null;
        }

        $done = (bool) ($request['done'] ?? false);
        $colorBackground = (bool) ($request['colorBackground'] ?? true);
        // -1 means uncapped; anything <1 (0, other negatives) normalizes back to -1 rather than
        // being rejected — there's no such thing as a column that holds zero tasks — matching
        // clampColumnCap's client-side twin (storage.js).
        $requestedCap = (int) ($request['cap'] ?? -1);
        $cap = $requestedCap < 1 ? -1 : $requestedCap;

        $stmt = $this->db->prepare('UPDATE "Columns" SET "Name" = :name, "Done" = :done, "Color" = :color, "ColorBackground" = :colorBackground, "Order" = :order, "Cap" = :cap WHERE "Id" = :id');
        $stmt->execute([
            'name' => $request['name'] ?? '', 'done' => (int) $done,
            'color' => $request['color'] ?? null, 'colorBackground' => (int) $colorBackground, 'order' => (int) ($request['order'] ?? 0), 'cap' => $cap, 'id' => $columnId,
        ]);

        return ['id' => $columnId, 'name' => $request['name'] ?? '', 'done' => $done, 'color' => $request['color'] ?? null, 'colorBackground' => $colorBackground, 'order' => (int) ($request['order'] ?? 0), 'cap' => $cap];
    }

    // ARCHITECTURE-REVIEW.md finding 3.1: unlink ParentTaskId -> delete TaskDependencies -> delete
    // Tasks -> delete Column used to run as four separately auto-committed statements — a
    // mid-sequence failure (e.g. the TaskDependencies delete) left orphaned rows referencing tasks
    // that either still exist with a dangling dependency, or were about to be deleted, either way an
    // inconsistent state no retry could cleanly recover from.
    public function delete(string $projectId, string $columnId): bool
    {
        $this->db->beginTransaction();
        try {
            $result = $this->deleteInTransaction($projectId, $columnId);
            $this->db->commit();
            return $result;
        } catch (\Throwable $e) {
            if ($this->db->inTransaction()) {
                $this->db->rollBack();
            }
            throw $e;
        }
    }

    private function deleteInTransaction(string $projectId, string $columnId): bool
    {
        $stmt = $this->db->prepare('SELECT 1 FROM "Columns" WHERE "Id" = :id AND "ProjectId" = :pid');
        $stmt->execute(['id' => $columnId, 'pid' => $projectId]);
        if ($stmt->fetch() === false) {
            return false;
        }

        // Tasks.ColumnId is a Restrict FK — deleting a column that still holds tasks would otherwise
        // fail at the DB level. Mirrors mutations.js's deleteColumn: every task in the column is
        // deleted outright (not reassigned elsewhere), with the same cleanup TasksController.delete
        // does per task — sub-tasks orphaned back to top-level (ParentTaskId is also Restrict),
        // dependency rows removed on both sides (DependsOnTaskId is Restrict too). Document/Risk/
        // Decision.TaskId and Tasks.AssigneeId are already SetNull, so those clear themselves.
        $stmt = $this->db->prepare('SELECT "Id" FROM "Tasks" WHERE "ColumnId" = :cid');
        $stmt->execute(['cid' => $columnId]);
        $taskIds = $stmt->fetchAll(PDO::FETCH_COLUMN);

        if ($taskIds !== []) {
            $placeholders = implode(',', array_fill(0, count($taskIds), '?'));

            $this->db->prepare(
                "UPDATE \"Tasks\" SET \"ParentTaskId\" = NULL WHERE \"ParentTaskId\" IN ($placeholders) AND \"Id\" NOT IN ($placeholders)"
            )->execute([...$taskIds, ...$taskIds]);

            $this->db->prepare(
                "DELETE FROM \"TaskDependencies\" WHERE \"TaskId\" IN ($placeholders) OR \"DependsOnTaskId\" IN ($placeholders)"
            )->execute([...$taskIds, ...$taskIds]);

            $this->db->prepare("DELETE FROM \"Tasks\" WHERE \"Id\" IN ($placeholders)")->execute($taskIds);
        }

        $this->db->prepare('DELETE FROM "Columns" WHERE "Id" = :id')->execute(['id' => $columnId]);
        return true;
    }
}
